lib/muchos/config/base.py

```diff
@@ -480,12 +505,26 @@ def checksum_ver(self, software, version):
                         )
 
         key = "{0}:{1}".format(software, version)
+
         if key not in self.checksums_d:
-            exit(
-                "ERROR - Failed to find checksums for {} {} in {}".format(
-                    software, version, self.checksums_path
+            local_tarball_path = self.get_local_tarball_path(software)
+            if local_tarball_path is None or not exists(local_tarball_path):
+                exit(
+                    "ERROR - Failed to find either a valid checksum in {}, "
+                    "or a local tarball to upload for {} {}.".format(
+                        self.checksums_path, software, version
+                    )
                 )
-            )
+            else:
+                # compute and use the checksum for local tarball
+                local_tarball_sha512 = sha512()
+                with open(local_tarball_path, "rb") as tarball_contents:
+                    file_buffer = tarball_contents.read(65536)
+                    while len(file_buffer) > 0:
+                        local_tarball_sha512.update(file_buffer)
+                        file_buffer = tarball_contents.read(65536)
+                return f"sha512:{local_tarball_sha512.hexdigest()}"
+
```

Comment:
I like the idea of a local tarball that is uploaded, but we still want to verify the local tarball isn't corrupt or malicious, so we shouldn't just recompute the checksum, but validate it locally before shipping it off to the cluster.

Reply:
@ctubbsii - this code path is only used when there is no checksum available (in conf/checksums) to compare against. Do you have any suggestions on how to accomplish what you suggested? I guess we could also require a .sha512 file to be placed in the conf/upload folder (for non-SNAPSHOT tarballs) and use that?

Reply:
Thinking through your comments in #428 I think we may be able to use checksums from conf/checksums (if available) to do this. I'll work through some changes and circle back on these (somewhat related) PRs.

Reply:
I'm just thinking about how fluo-uno works: if the file is already downloaded, it won't download it again. However, it does always verify the local file's checksum before using it.

If we merely verify a checksum we've computed ourselves, we're not doing anything to ensure the artifacts themselves are legitimate.

Reply:
Please review my subsequent commit. I believe it addresses the previous comments. To summarize, these changes now offer added flexibility, without sacrificing security:
* For archive releases, we can now use local tarballs from conf/upload, **only** if they have either a checksum in conf/checksums, or a matching .sha512 file present in conf/upload.
  * I tested this with samples of archived Hadoop, Zookeeper and Spark releases which do not have entries in conf/checksums - they are all good.
  * For good measure, I also tested a case where I placed the released Hadoop 3.1.4 tarball alone in the conf/upload folder (without a .sha512 file) and validated that Mucho correctly used the matching checksum from conf/checksums.
* For SNAPSHOT tarballs, we will compute a checksum locally and use that on the Ansible side to ensure that the tarball was not corrupted in transit. This is an enhancement over what we have currently in main.

By the way, in neither my original commit, nor in this one, will the code download a tarball if it is already present and the checksum matches. This behavior is also status quo from what is in main.